torchtune/datasets/_packed.py

```diff
@@ -0,0 +1,274 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Any, Dict, List, Optional
+
+import torch
+from torch.nn import functional as F
+
+from torch.utils.data import Dataset
+from torchtune.data import CROSS_ENTROPY_IGNORE_IDX
+from torchtune.utils import get_world_size_and_rank
+from tqdm import tqdm
+
+
+class PackedDataset(Dataset):
+    """
+    Performs greedy sample packing on a provided dataset. This is done as a single
+    preprocessing step before training begins. Shuffling is done outside of this
+    class on packed samples with a ``Sampler`` as part of the dataloader. Currently,
+    this only supports in-memory map-style datasets.
+
+    The class loads, tokenizes, and packs examples on initialization - no tokenization is done during training.
+
+    The general flow on initialization is: load tokenized sample -> add to buffer ->
+    when buffer is long enough, add to ``self.packs``.
+
+    During training, returns self.packs[idx] as input, label, attention mask, and
+    position ids. The attention mask is a lower triangular block mask to prevent
+    samples from cross-attending within a pack. The position ids indicate the position
+    of each token relative to its sample within a pack. These are all padded to max
+    sequence length, so a batch-wise collator is not needed.
+
+    A packed sample is made up of individual smaller sequence length samples jammed together
+    within ``max_seq_len``. For example, if max_seq_len is 6 and there are varied
+    length samples::
+
+        tokens = [
+            [S1, S1, S1, S2, S2, pad],
+            [S3, S3, S4, S4, pad, pad],
+            ...,
+        ]
+
+    To prevent cross-contamination, the following mask would be returned for the
+    first pack in the example::
+
+        mask = [
+            [1, 0, 0, 0, 0, 0],
+            [1, 1, 0, 0, 0, 0],
+            [1, 1, 1, 0, 0, 0],
+            [0, 0, 0, 1, 0, 0],
+            [0, 0, 0, 1, 1, 0],
+            [0, 0, 0, 0, 0, 1],
+        ]
+
+    The position ids would be::
+
+        input_pos = [
+            [0, 1, 2, 0, 1, 2],
+            [0, 1, 0, 1, 2, 3],
+            ...,
+        ]
+
+    The identity matrix is used in the mask for pad tokens instead of a causal mask.
+    For position ids for pad tokens, we simply continue to increment from the previous
+    sample normally.
+
+    Args:
+        ds (Dataset): dataset to sample pack. This should return a dictionary with field
+            "tokens" and "labels" containing the tokenized and label samples.
+        max_seq_len (int): Maximum number of tokens to pack
+        max_packs (Optional[int]): maximum number of packs. Default is None, which will create as many
```

Comment:
Do we have the appropriate error checking here and do we really need to expose this? What if I specify max_packs as something silly like 1 or 2, but the dataset packed into 2 samples isn't possible because it exceeds max_seq_len?

Reply:
it won't pack the entire dataset into 2 packs, it will just create 2 packs with whatever samples possible and then drop the rest.

`max_packs` may become more relevant for iterable datasets, but maybe it's not as useful right now. Will revisit this.